{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy as scp\n",
    "from sklearn import preprocessing\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method for finding entry indexex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method to find subseq in seq\n",
    "\n",
    "def knutt_morris(txt, pattern):\n",
    "    '''\n",
    "    Knutt-Morris-Pratt algorithm for finding subsequence in sequence.\n",
    "    Example: K_M_P(pattern, txt),\n",
    "        pattern = subsequence to find,\n",
    "        txt = sequence\n",
    "    return:\n",
    "        list of all index, where is subseq was found.\n",
    "        '''\n",
    "    result = []\n",
    "    \n",
    "    len_pat = len(pattern) \n",
    "    len_txt = len(txt) \n",
    "  \n",
    "    pref_suff = [0]*len_pat \n",
    "    j = 0 \n",
    "      \n",
    "    compute_pref_suff(pattern, len_pat, pref_suff) \n",
    "  \n",
    "    i = 0\n",
    "    while i < len_txt:\n",
    "        if pattern[j] == txt[i]: \n",
    "            i += 1\n",
    "            j += 1\n",
    "  \n",
    "        if j == len_pat: \n",
    "            result.append((i-j)) \n",
    "            j = pref_suff[j-1] \n",
    "            \n",
    "        elif i < len_txt and pattern[j] != txt[i]: \n",
    "            if j != 0: \n",
    "                j = pref_suff[j-1] \n",
    "            else: \n",
    "                i += 1\n",
    "                \n",
    "    return result\n",
    "  \n",
    "def compute_pref_suff(pattern, len_pat, pref_suff):\n",
    "    '''Pref_suff func for K-M-P algorithm'''\n",
    "    length = 0\n",
    "    i = 1\n",
    "  \n",
    "    # the loop calculates lps[i] for i = 1 to len_pat-1 \n",
    "    while i < len_pat: \n",
    "        if pattern[i]== pattern[length]: \n",
    "            length += 1\n",
    "            pref_suff[i] = length\n",
    "            i += 1\n",
    "        else: \n",
    "            if length != 0: \n",
    "                length = pref_suff[length-1] \n",
    "            else: \n",
    "                pref_suff[i] = 0\n",
    "                i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "prot_records = []\n",
    "for file in data_files:\n",
    "    records_organism = list(SeqIO.parse(file, 'fasta'))\n",
    "    prot_records.append(records_organism)\n",
    "    \n",
    "#print(records[11])\n",
    "\n",
    "#records - list of lists with records for each organism\n",
    "\n",
    "prot_records[0]\n",
    "\n",
    "\n",
    "index = 0\n",
    "for i in range(len(prot_records)):\n",
    "    org = organisms[i]\n",
    "    org_prot_rec = prot_records[i]\n",
    "    for protein in org_prot_rec:\n",
    "        adding_row = []\n",
    "        adding_row.append(org)\n",
    "        adding_row.append(protein.description)\n",
    "        adding_row += finding_frequency(str(protein.seq))\n",
    "        #print(adding_row)\n",
    "        Proteins_data.loc[index] = adding_row\n",
    "        index += 1\n",
    "    print(str(i+1) + ' of ' + str(len(prot_records)) + ' organisms are preprocessed. Party time!')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_test_file = SeqIO.parse('../data/Campylobacter_jejuni.fasta', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "record = list(fasta_test_file)[0]\n",
    "\n",
    "\n",
    "def read_seqIO(SeqRecord):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test string analyzes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "string = 'MAWRSGLCETDSRTLKQFLQEECMWKLVGKSRKHREYRAVACRSTIFSPEDDSSCILCQL\\\n",
    "LLLYRDGEWIICFCCNGRYQGHYGVNHVHRRRRRICHLPTLYQLSFGGPLGPASIDFLPS\\\n",
    "FSQVTSSMTCDGITPDVIYEVCMLVPQDEAKRILVKGHGAMDLTCQKAVTLGGAGAWLLP\\\n",
    "RPEGYTLFFYILCYDLFTSCGNRCDIPSMTRLMAAATACGQAGCSFCTDHEGHVDPTGNY\\\n",
    "VGCTPDMGRCLCYVPCGPMTQSLIHNEEPATFFCESDDAKYLCAVGSKTAAQVTLGDGLD\\\n",
    "YHIGVKDSEGRWLPVKTDVWDLVKVEEPVSRMIVCSCPVLKNLVH'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First attempt with classes, was considered not so good but saved for later\n",
    "# You can pass it\n",
    "\n",
    "class K_mer_str:\n",
    "    '''\n",
    "    Class for K-mer analyzes\n",
    "    '''\n",
    "    \n",
    "    def __init__(string, k):\n",
    "        self.string = string\n",
    "        self.k = k\n",
    "        self.vector = []\n",
    "        \n",
    "    def entry_index_vector(self, subseq):\n",
    "        n = none\n",
    "        pass\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For example we will try to find all needed vectors for 4-mers\n",
    "\n",
    "import itertools\n",
    "\n",
    "\n",
    "# Amino acid data\n",
    "\n",
    "amino_dict = { 'ALA': 'A', 'ARG': 'R',\n",
    "                  'ASN': 'N', 'ASP': 'D',\n",
    "                  'CYS': 'C', 'GLN': 'Q',\n",
    "                  'GLU': 'E', 'GLY': 'G',\n",
    "                  'HIS': 'H', 'ILE': 'I',\n",
    "                  'LEU': 'L', 'LYS': 'K',\n",
    "                  'MET': 'M', 'PHE': 'F',\n",
    "                  'PRO': 'P', 'SER': 'S',\n",
    "                  'THR': 'T', 'TRP': 'W',\n",
    "                  'TYR': 'Y', 'VAL': 'V' }\n",
    "\n",
    "list_amino = list(amino_dict.values())\n",
    "#list_amino\n",
    "amino_string = ''.join(x for x in list_amino)\n",
    "\n",
    "amino_string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make all posible combinations\n",
    "# K-mer = 4\n",
    "\n",
    "subseq_iter = itertools.product(amino_string, repeat = 4)\n",
    "\n",
    "subseq_list = list(subseq_iter)\n",
    "print(subseq_list[0]) #1\n",
    "\n",
    "for i in range(len(subseq_list)):\n",
    "    tup = subseq_list[i]\n",
    "    subseq_list[i] = ''.join(tup)\n",
    "\n",
    "print(subseq_list[0]) #2\n",
    "print(len(subseq_list)) #3\n",
    "\n",
    "# ok, we are ready"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making functions\n",
    "\n",
    "def vectorizing_single_subseq(string, subseq, n, k_mer_num,\n",
    "                              knutt_morris:callable):\n",
    "    '''\n",
    "    finding subseq's entry index vector for subseq for some string\n",
    "    return vector\n",
    "    '''\n",
    "    \n",
    "    #n = len(string)\n",
    "    #m = len(subseq)\n",
    "    \n",
    "    vector = [0 for k in range(n-k_mer_num+1)]\n",
    "    \n",
    "    list_entry = knutt_morris(string, subseq)\n",
    "    \n",
    "    for x in list_entry:\n",
    "        vector[x] = 1\n",
    "        \n",
    "    return np.array(vector)\n",
    "\n",
    "def vectorizing_all_subseq(string, subseq_list,\n",
    "                           vectorizing_single_subseq:callable):\n",
    "    '''\n",
    "    finding subseq's entry index vector for all subseqs from subseq_list\n",
    "    !!! return dict: subseq -> vector\n",
    "    '''\n",
    "    n = len(string)\n",
    "    k_mer_num = len(subseq_list[0])\n",
    "    \n",
    "    vector_scaled_list = []\n",
    "    #matrix = []\n",
    "    \n",
    "    for subseq in subseq_list:\n",
    "        vector = vectorizing_single_subseq(string, subseq, n, k_mer_num, knutt_morris)\n",
    "        vector = vector.reshape((-1, 1))\n",
    "        \n",
    "        scaler = preprocessing.StandardScaler() #  SKlearn\n",
    "        \n",
    "        vector_scaled = scaler.fit_transform(vector)\n",
    "        del vector\n",
    "        vector_scaled_list.append(vector_scaled)\n",
    "        del vector_scaled\n",
    "    \n",
    "    return dict(zip(subseq_list, vector_scaled_list))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "prot_records = []\n",
    "for file in data_files:\n",
    "    records_organism = list(SeqIO.parse(file, 'fasta'))\n",
    "    prot_records.append(records_organism)\n",
    "    \n",
    "#print(records[11])\n",
    "\n",
    "#records - list of lists with records for each organism\n",
    "\n",
    "prot_records[0]\n",
    "\n",
    "\n",
    "index = 0\n",
    "for i in range(len(prot_records)):\n",
    "    org = organisms[i]\n",
    "    org_prot_rec = prot_records[i]\n",
    "    for protein in org_prot_rec:\n",
    "        adding_row = []\n",
    "        adding_row.append(org)\n",
    "        adding_row.append(protein.description)\n",
    "        adding_row += finding_frequency(str(protein.seq))\n",
    "        #print(adding_row)\n",
    "        Proteins_data.loc[index] = adding_row\n",
    "        index += 1\n",
    "    print(str(i+1) + ' of ' + str(len(prot_records)) + ' organisms are preprocessed. Party time!')\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
