{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import shutil\n",
    "import os\n",
    "from joblib import dump, load\n",
    "import urllib3\n",
    "import certifi\n",
    "from Bio import SeqIO\n",
    "import Bio\n",
    "from glob import glob\n",
    "import json\n",
    "from IPython import display\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "from collections import OrderedDict\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HIV properties\n",
    "\n",
    "patients = [\"p{}\".format(i) for i in range(1,12)]\n",
    "hiv_regions = [\"V3\", \"PR\", \"psi\", \"vpr\", \"vpu\", \"p1\", \"p2\", \"p6\", \"p7\", \"p15\", \"p17\", \"RRE\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_hivevo_haplotype(patient, folder, hiv_region):\n",
    "    '''\n",
    "    Downloading haplotypes for region for patient\n",
    "    '''\n",
    "    global hi\n",
    "    http = urllib3.PoolManager(cert_reqs='CERT_REQUIRED',\n",
    "                               ca_certs=certifi.where())\n",
    "    \n",
    "    api = \"https://hiv.biozentrum.unibas.ch/api/data/haplotypes/\"\n",
    "    \n",
    "    url = \"/\".join((api, patient, hiv_region))\n",
    "\n",
    "    if not os.path.isdir(folder):\n",
    "        os.mkdir(folder)\n",
    "\n",
    "    file_name = folder + \"_\".join((\"hivevo\", patient, region)) + \".fasta\"\n",
    "\n",
    "    with http.request('GET', url, preload_content=False) as res, open(file_name, 'wb') as out_file:\n",
    "\n",
    "        shutil.copyfileobj(res, out_file)\n",
    "\n",
    "        \n",
    "\n",
    "folder = \"../data/regions/\"\n",
    "for patient in patients:\n",
    "    for region in hiv_regions:\n",
    "        download_hivevo_haplotype(patient, region, folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_hivevo_references(folder):\n",
    "    '''\n",
    "    Downloading references for each patient and storing them in dir folder/references\n",
    "    '''\n",
    "    global patients    \n",
    "    \n",
    "    http = urllib3.PoolManager(cert_reqs='CERT_REQUIRED',\n",
    "                               ca_certs=certifi.where())\n",
    "    \n",
    "    if not os.path.isdir(folder):\n",
    "        os.mkdir(folder)\n",
    "    \n",
    "    for patient in patients:\n",
    "        api = \"https://hiv.biozentrum.unibas.ch/api/data/referenceSequence\"\n",
    "        url = \"/\".join((api, patient))\n",
    "        file_name = folder + \"_\".join((\"hivevo\", \"reference\", patient)) + \".fasta\"\n",
    "        \n",
    "        with http.request('GET', url, preload_content=False) as res, open(file_name, 'wb') as out_file:\n",
    "            shutil.copyfileobj(res, out_file)\n",
    "\n",
    "\n",
    "folder = \"data/references/\"\n",
    "download_hivevo_references(folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extracting_region_from_reference(region, reference_path, folder):\n",
    "    '''\n",
    "    Extracting special region from reference\n",
    "    '''\n",
    "    \n",
    "    if not os.path.isdir(folder+region):\n",
    "        os.mkdir(folder+region)\n",
    "    \n",
    "    with open(reference_path) as f:\n",
    "        reference_info = json.load(f)\n",
    "    for reg in reference_info['features']:\n",
    "        if reg['name'] == region:\n",
    "            loc = reg['location'][0]\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "    patient = re.search(r'p[\\d]*', reference_path)[0]\n",
    "    #print(patient)\n",
    "    sequence = reference_info['seq'][loc[0]:loc[1]]\n",
    "    #print(sequence)\n",
    "    res = r'/'+re.search(r'[\\w]*\\.fasta', reference_path)[0].replace(patient, \"_\".join((patient, region)))\n",
    "    \n",
    "    with open(folder+region+res, 'w') as write_file:\n",
    "        write_file.write('>'+reference_info['description'].replace('genomewide', 'region='+region)+'\\n')\n",
    "        write_file.write(sequence)\n",
    "\n",
    "\n",
    "# UNCOMMENT THIS IF YOU ARE BRAVE        \n",
    "#folder = 'data/references/'\n",
    "#region = 'V3'\n",
    "#ref_path = 'data/references/hivevo_reference_p4.fasta'\n",
    "#extracting_region_from_reference(region, ref_path, folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def json2fasta(folder, path_json):\n",
    "    '''\n",
    "    Converting json to fasta\n",
    "    WARNING!!! works not as properly as it should. be carefull\n",
    "    '''\n",
    "    \n",
    "    if not os.path.isdir(folder+'fasta'):\n",
    "        os.mkdir(folder+'fasta')\n",
    "        \n",
    "    with open(path_json) as f:\n",
    "        json_file = json.load(f)\n",
    "    \n",
    "    path = path_json.replace('data/', 'data/fasta/')   \n",
    "    with open(path, \"w\") as fasta_file:\n",
    "        for obj in json_file:\n",
    "            line_1 = '>' + obj['name'] + '\\n'\n",
    "            line_2 = obj['sequence'] + '\\n'\n",
    "            lines = [line_1, line_2]\n",
    "            fasta_file.writelines(lines)\n",
    "            \n",
    "# UNCOMMENT THIS IF YOU ARE BRAVE\n",
    "#json2fasta('data/','data/hivevo_p4_V3.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_ref_reg2fasta(path_fasta, path_ref_region):\n",
    "    '''\n",
    "    Adding exact region from reference file to fasta with haplotypes\n",
    "    '''\n",
    "    with open(path_fasta, 'a') as fasta, open(path_ref_region) as ref:\n",
    "        for line in ref:\n",
    "            #print(line)\n",
    "            fasta.write(line)\n",
    "            \n",
    "#add_ref_reg2fasta('data/fasta/hivevo_p4_V3.fasta', 'data/references/V3/hivevo_reference_p4_V3.fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
